src/metabase/driver/clickhouse.clj

```diff
@@ -112,6 +119,72 @@
            :semantic-version {:major (.getInt rset 2)
                               :minor (.getInt rset 3)}})))))
 
+(defmethod driver/upload-type->database-type :clickhouse
+  [_driver upload-type]
+  (case upload-type
+    ::upload/varchar-255              "Nullable(String)"
+    ::upload/text                     "Nullable(String)"
+    ::upload/int                      "Nullable(Int64)"
+    ::upload/float                    "Nullable(Float64)"
+    ::upload/boolean                  "Nullable(Boolean)"
+    ::upload/date                     "Nullable(Date32)"
+    ::upload/datetime                 "Nullable(DateTime64(3))"
+    ;; FIXME: should be `Nullable(DateTime64(3))`
+    ::upload/offset-datetime          nil))
+
+(defmethod driver/table-name-length-limit :clickhouse
+  [_driver]
+  ;; FIXME: This is a lie because you're really limited by a filesystems' limits, because Clickhouse uses
+  ;; filenames as table/column names. But its an approximation
+  206)
+
+(defn- quote-name [s]
+  (let [parts (str/split (name s) #"\.")]
+    (str/join "." (map #(str "`" % "`") parts))))
+
+(defn- create-table!-sql
+  [driver table-name column-definitions & {:keys [primary-key]}]
+  (str/join "\n"
+            [(first (sql/format {:create-table (keyword table-name)
+                                 :with-columns (mapv (fn [[name type-spec]]
+                                                       (vec (cons name [[:raw type-spec]])))
+                                                     column-definitions)}
+                                :quoted true
+                                :dialect (sql.qp/quote-style driver)))
+             "ENGINE = MergeTree"
+             (format "PRIMARY KEY (%s)" (str/join ", " (map quote-name primary-key)))
+             "ORDER BY ()"]))
+
+(defmethod driver/create-table! :clickhouse
+  [driver db-id table-name column-definitions & {:keys [primary-key]}]
+  (let [sql (create-table!-sql driver table-name column-definitions :primary-key primary-key)]
+    (qp.writeback/execute-write-sql! db-id sql)))
+
+(defmethod driver/insert-into! :clickhouse
+  [driver db-id table-name column-names values]
+  (when (seq values)
+    (sql-jdbc.execute/do-with-connection-with-options
+     driver
+     db-id
+     {:write? true}
+     (fn [^java.sql.Connection conn]
+       (let [sql (format "insert into %s (%s)" (quote-name table-name) (str/join ", " (map quote-name column-names)))]
+         (with-open [ps (.prepareStatement conn sql)]
+           (doseq [row values]
+             (when (seq row)
+               (doseq [[idx v] (map-indexed (fn [x y] [(inc x) y]) row)]
+                 (condp isa? (type v)
```

Comment:
Probably not a bottleneck, but making a protocol like `(ps-set! v ps idx)` could remove the reflection of `type` and the O(n) branching of `condp`.

Reply:
Yes, how do you imagine this being implemented exactly? The tricky part is this multimethod doesn't know the type of each column without using reflection, and values can be `nil`. I'm imagining this:
- for each value, call `(ps-set! v ps idx)`
- `px-set!` looks up the method to use for the `idx` in a volatile map, e.g. `setString` etc. If the map contains a method for the `idx`, it uses it. Otherwise calculate the method using a similar `condp` expression and save it to the map under the key `idx`, and execute the method. 

That way we'll only use reflection once for every column.

Reply:
Maybe I'm missing something, but I was thinking we'd just implement the protocol directly on each of these java types for `v`, eschewing any reflection or map look-up. I think you can just use `Object` for the fallback case.

Reply:
Right, anyway this seems pretty inconsequential compared to all the slow stuff we're doing on the Metabase side. I think I'd prefer to keep it simple and not change this `condp` for now